### PR TITLE
Add INBC SOTA update, download-only and no-download modes

### PR DESCRIPTION
**Pull request by tsirlapu**
_The PR review is to check for sustainability and correctness.  Sustainability is actually more business critical as correctness is largely tested into the code over time.   Its useful to keep in mind that SW often outlives the HW it was written for and engineers move from job to job so it is critical that code developed for Intel be supportable across many years.  It is up to the submitter and reviewer to look at the code from a perspective of what if we have to debug this 3 years from now after the author is no longer available and defect databases have been lost.  Yes, that happens all the time when we are working with time scales of more than 2 years.  When reviewing your code it is important to look at it from this perspective._

Author Mandatory (to be filled by PR Author/Submitter)
------------------------------------------------------
- Developer who submits the Pull Request for merge is required to mark the checklist below as applicable for the PR changes submitted.  
- Those checklist items which are not marked are considered as not applicable for the PR change.  
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the developers for Inner Source Development Model (ISDM) compliance

### PULL DESCRIPTION
_Provide a 1-2 line brief overview of the changes submitted through the Pull Request..._
SOTA to support update, no_download and download only modes

### REFERENCES
Reference URL for issue tracking (JIRA/HSD/Github): https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=530032
- [x] **_Added label to the Pull Request following the template: ISDM\_\<Complexity>\*_** \
	Note-1: Depending on complexity of code changes, use the suitable word for complexity: Low/Medium/High \
	Example: PR for Slim boot loader project with medium complexity can have the label as: ISDM_Medium	
- [x] Added label to the Pull Request for easier discoverability and search
- [x] RTC or HSD number will be included in final merge. HSD must always be included if available.
- [x] Changelogs are updated (or N/A if not customer visible)
- [N/A] inbm/log_changes.txt and inbm-vision/log_changes.txt are updated for potentially Validation-breaking log changes (or N/A if none)

### CODE MAINTAINABILITY
- [ ] **_Commit Message meets guidelines as indicated in the URL\*_**
	- (https://github.com/edgexfoundry/edgex-go/blob/main/.github/Contributing.md)
- [ ] **_Every commit is a single defect fix and does not mix feature addition or changes\*_**
- [ ] Added required new tests relevant to the changes
	- [ ] PR contains URL links to functional tests executed with the new tests 
- [ ] Updated Documentation as relevant to the changes
- [ ] Updated Build steps/commands changes as relevant
- [ ] PR change contains code related to security
- [ ] PR introduces changes that breaks compatibility with other modules (If YES, please provide description)
- [ ] Specific instructions or information for code reviewers (If any):
- [ ] Run 'go fmt' or autopep8 as applicable.
- [ ] New/modified methods and functions should have type annotations on signatures as applicable
- [ ] New/modified methods must have appropriate doc strings (language dependent)

### APPLICATION SPECIFIC
- [ ] Does PR change default config files under /etc? If so, will application still work after an upgrade that leaves /etc alone, like a Mender upgrade?
- [ ] Is cloud UI changed? If so, are cloud definition files updated?

Maintainer Mandatory (to be filled by PR Reviewer/Approving Maintainer)
-----------------------------------------------------------------------
- Maintainer who approves the Pull Request for merge is required to mark the checklist below as appropriate for the PR change reviewed as key proof of attestation indicating reasons for merge. 
- Those checklist items which are not marked are considered as not applicable for the PR change. 
- Items marked with an asterisk suffix are mandatory items to check and if not marked will be treated as non-compliant pull requests by the maintainers for ISDM compliance.

### QUALITY CHECKS
- [ ] Architectural and Design Fit
- [ ] **_Quality of code (At least one should be checked as applicable)\*_**
	- [ ] Commit Message meets guidelines
	- [ ] PR changes adhere to industry practices and standards
	- [ ] Error and exception code paths implemented correctly
	- [ ] Code reviewed for domain or language specific anti-patterns
	- [ ] Code is adequately commented
	- [ ] Code copyright is correct
	- [ ] Confusing logic is explained in comments
	- [ ] Commit comment can be used to design a new test case for the changes
- [ ] **_Test coverage shows adequate coverage with required CI functional tests pass on all supported platforms\*_**
- [ ] **_Static code scan report shows zero critical issues\*_**
- [ ] Integration tests are passing

### CODE REVIEW IMPACT
- Summary of Defects Detected in Code Review: **\<%P1*xx,P2*xx,P3*xx,P4*xx%>** \
Note P1/P2/P3/P4 denotes severity of defects found (Showstopper/High/Medium/Low) and xx denotes number of defects found

### SECURITY CHECKS
Please check if your PR fulfills the following requirements:

- [ ] Follow best practices when handling primitive data types
- [ ] Configure minimal permissions when opening pipes and ports
- [ ] Check contents within input structures are valid before use
- [ ] All forms of input validated
- [ ] Avoid inter-process race conditions
- [ ] Error and exception handling implemented
- [ ] Defend against Canonical Representation Issues - Any paths utilized?
- [ ] Follow 'secure by default' - Any configuration values added
- [ ] Fail safe - Any failure scenarios?
- [ ] Clean up temporary files - Any temporary files being used?

# _Code must act as a teacher for future developers_

### CODE REVIEW

File: inbm/dispatcher-agent/dispatcher/sota/os_upgrader.py
Comment (gblewis1) [inline inbm/dispatcher-agent/dispatcher/sota/os_upgrader.py:73]: --no-download?

File: inbm/dispatcher-agent/dispatcher/sota/os_upgrader.py
Comment (gblewis1) [inline inbm/dispatcher-agent/dispatcher/sota/os_upgrader.py:84]: --download-only --fix-missing (and remove apt-get upgrade --fix-missing above)

**Review by gblewis1 — CHANGES_REQUESTED** [review]
see comments

File: inbc-program/README.md
Comment (nmgaston) [inline inbc-program/README.md:106]: 'SOTA in this mode' doesn't need to be stated.  That's implied.  Direct statements are better.  'Performs the software update'. 'Retrieves and installs packages', 'Retrieves the packages only (will not unpack or install).

File: inbc-program/README.md
Comment (nmgaston) [inline inbc-program/README.md:124]: Need to put the --command in the correct brackets for whether it is required or optional.  { } is required and [ ] is optional.

File: inbc-program/inbc/parser.py
Comment (nmgaston) [inline inbc-program/inbc/parser.py:107]: As Gavin stated, this should be mode instead of command.

File: inbc-program/README.md
Comment (nmgaston) [reply inbc-program/README.md:129]: I believe she did it this way, because in the manifest it has <cmd as the tag and we need it to be backwards compatible.  The word 'mode' makes more sense in this case, but I can see using command since that is what the manifest states already.

File: inbm/common-python-config/pyproject.toml
Comment (nmgaston) [reply inbm/common-python-config/pyproject.toml:44]: agree

File: inbm/dispatcher-agent/dispatcher/dispatcher_class.py
Comment (nmgaston) [inline inbm/dispatcher-agent/dispatcher/dispatcher_class.py:857]: The switching back and forth between the turns mode and command is confusing for the same thing.  Since we are using command for the rollback, we probably should use mode throughout even if it's in the manifest with the 'cmd' tag.

File: inbc-program/README.md
Comment (nmgaston) [reply inbc-program/README.md:129]: But now that I see below that we are using cmd for the rollback, I agree that we should be using mode everywhere even if this get in the manifest under the cmd tag.

Comment (gblewis1) [reply inbc-program/README.md:129]: @nmgaston I was thinking `inbc sota --command update --mode download-only` and that `--command update` was implied (default).

Comment (gblewis1) [reply inbc-program/README.md:129]: (as the previous invocation was `inbc sota` by itself)

File: inbc-program/README.md
Comment (tsirlapu) [reply inbc-program/README.md:134]: @gblewis1 I'll change that

File: inbc-program/README.md
Comment (tsirlapu) [reply inbc-program/README.md:129]: @gblewis1  as mentioned in the story and also the email change I'll stick to inbc sota --mode download-only

File: inbm/dispatcher-agent/dispatcher/sota/os_updater.py
Comment (gblewis1) [inline inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: Check the existing method that does the full update to see how we do it in the 'full' mode. Also consider whether you want to reboot after no-download/download-only and how these modes interact with logging the update. Finally, if download-only is run and you're not rebooting, make sure we aren't going to health check/roll back if there's a reboot (since we only ran a download), at least not until we apply the update later in no-download mode.

Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: @gblewis1 the reboot part is handled in other update by yeng Liong. we are not looking at it in this story

Comment (gblewis1) [reply inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: > @gblewis1 the reboot part is handled in other update by yeng Liong. we are not looking at it in this story

That's not how I read the story. https://rtc.intel.com/ccm0006001/web/projects/SSEA-%20IOTG%20PED%20Team#action=com.ibm.team.workitem.viewWorkItem&id=530032 has a comment saynig  'Do ensure that if the sita cmd is a "download-only", the device should NOT go through a reboot' and there is a child story with title '[LP] Do ensure that if the sota cmd is a "download-only", the device should NOT go through a reboot'.  @pillaiabhishek has this changed?

Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: @gblewis1 yes it is @pillaiabhishek who got that part assigned to Yeng Liong, he is currently working on that MR. This MR does not handle reboot functionality.

Comment (pillaiabhishek) [reply inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: I dont see a reason why a device needs to go through a reboot if in a "download-only" sota cmd.
Any concerns here @tsirlapu @gblewis1 ?

Comment (pillaiabhishek) [reply inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: @tsirlapu, Yeng Liong needs to work on adding a --reboot option in the SOTA cmd.
But even without that, in this PR we need to ensure that if a user sends a SOTA "download-only" cmd, the device does not reboot.

Below is expected from the PR:
1. inbc sota _-> Perform download and install. Reboots the platform._
2. inbc sota --command update _-> Perform download and install. Reboots the platform._
3. inbc sota --command update --mode download-only _-> Perform download only. No platform reboot._
4. inbc sota --command update --mode no-download _-> Perform install only. Reboots the platform._

Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: @pillaiabhishek this is already discussed, This MR is without the reboot functionality, the next MR is with Reboot functionality,  here are the changes for reboot https://github.com/intel/intel-inb-manageability/pull/345

Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/os_updater.py:180]: @gblewis1 we are good, this MR does not cover reboot functionality changes, @pillaiabhishek just confirmed as discussed in our email chain "sounds good Teja. 
Do let Gavin know that Yeng Liong already has that covered." Changes of reboot covered in MR345

**Review by gblewis1 — CHANGES_REQUESTED** [review]
Looks good, let me know when it's tested and I'll approve and merge (if you have Natalie's approval as well).

File: inbc-program/inbc/parser.py
Comment (nmgaston) [inline inbc-program/inbc/parser.py:105]: Since the command is always set to update.  I question if we actually need to add that to INBC.  We can just hardcode that to update and if there ever is another option, then we can add it to INBC.

File: inbc-program/inbc/parser.py
Comment (nmgaston) [inline inbc-program/inbc/parser.py:296]: I don't think we need to add the update to command to INBC since it's always update.  I'd leave this hardcoded and remove the command from INBC as an option.

File: inbc-program/README.md
Comment (tsirlapu) [reply inbc-program/README.md]: @nmgaston Addressed

File: inbc-program/inbc/parser.py
Comment (tsirlapu) [reply inbc-program/inbc/parser.py:105]: @nmgaston Yes agree, but this is just a placeholder, Would you want me to remove this?

File: inbc-program/README.md
Comment (yengliong93) [inline inbc-program/README.md:132]: should it be `inbc sota --command update --mode download-only`?

File: inbc-program/README.md
Comment (yengliong93) [inline inbc-program/README.md:137]: should it be `inbc sota --command update --mode no-download`?

Comment (tsirlapu) [reply inbc-program/README.md:137]: @yengliong93 by default --command is update so don't have to use it is what Natalie suggested

These are examples, so you don't want to use the brackets. And since there are default values set, you won't need to show the command and mode unless it's different than the default.

inbc sota (don't need the update and full because they are the updates)
inbc sota --mode download-only
inbc sota --mode no-download

File: inbc-program/README.md
Comment (tsirlapu) [reply inbc-program/README.md:111]: @nmgaston fixed

File: inbc-program/inbc/parser.py
Comment (nmgaston) [reply inbc-program/inbc/parser.py:105]: I would.  I think it's confusing since there isn't any real option at this time.  If I saw that in someone else's software I would really question it.

File: inbc-program/README.md
Comment (nmgaston) [inline inbc-program/README.md:115]: can you add a space after "full", so it's consistent.

Comment (tsirlapu) [reply inbc-program/README.md]: @nmgaston done

File: inbm/dispatcher-agent/dispatcher/sota/sota.py
Comment (nmgaston) [inline inbm/dispatcher-agent/dispatcher/sota/sota.py:144]: should line 144 still be sota_cmd?

File: inbc-program/README.md
Comment (nmgaston) [inline inbc-program/README.md:114]: since we are removing --command as an INBC parameter.  We should remove it from the documentation in line 114

File: inbm/dispatcher-agent/dispatcher/sota/sota.py
Comment (gblewis1) [reply inbm/dispatcher-agent/dispatcher/sota/sota.py:144]: agreed, this should be sota_cmd

Comment (tsirlapu) [reply inbm/dispatcher-agent/dispatcher/sota/sota.py]: fixed

File: inbc-program/README.md
Comment (tsirlapu) [reply inbc-program/README.md]: fixed

**Review by nmgaston — APPROVED** [review]

**Review by gblewis1 — APPROVED** [review]

**Pull request merged**
